In [12]:
import sys
sys.path.append('../')
import pandas as pd
import numpy as np
from xgboost import XGBRegressor
from datetime import timedelta
import pickle

from utils import utils_scenario as utils, data_preparation as prep, data_process as proc

In [13]:
#read in processed measurement data and forecast data
measurement = pd.read_csv('./data/processed/last_measurement.csv')
measurement['datetime']= pd.to_datetime(measurement['datetime'],format='%Y-%m-%d %H:%M:%S')
forecast = pd.read_csv('./data/processed/last_forecast.csv')
forecast['f_date']= pd.to_datetime(forecast['f_date'],format='%Y-%m-%d %H:%M:%S')
# forecast['p_date']= pd.to_datetime(forecast['p_date'],format='%Y-%m-%d %H:%M:%S')

In [14]:
#prepare processed data into same format as training df 
def prepare_x_test(measurement, forecast, past_n_steps, pred_period):
    
    #concat past n steps from measurement 
    df = measurement.set_index('datetime')
    #drop am feature
#     df = df.drop(['am'], axis=1)
    df=proc.get_past_n_steps(df, past_n_steps)
    #calculate forecast_time
    df['forecast_time'] = df['present_time']+timedelta(hours=pred_period)

    #join forecast according to forecast time 
    forecast = forecast.set_index('f_date') 
    forecast = forecast.add_suffix('_forecast')
    df = pd.merge(df, forecast, how = 'left', left_on = 'forecast_time', right_on ='f_date')
    #add cos day
    df = proc.smooth_day_hour(df)
    #fillna
    df.fillna(value=0, inplace=True)
    df = df.iloc[:-past_n_steps]
    #keep the latest information
    df_out = df.iloc[-1:]
    return df_out

## Check if any columns are missing:

In [16]:
# #to change: make an assert error
# df= prepare_x_test(measurement, forecast, 48, 6)
# #load model columns 
# model_cols = pickle.load(open('model_cols_past_48.pkl', 'rb'))
# print('columns missing:', model_cols.difference(df.columns))
# #it's okay if am feature is missing - it's not included in the model

In [17]:
df

speed_t-48  cos_wind_dir_t-48  sin_wind_dir_t-48  temp_t-48  \
280       5.545           0.596844          -0.766326  28.508333   

     radiation_t-48  precip_t-48  season_t-48  speed_t-47  cos_wind_dir_t-47  \
280      590.366667          0.0            3    4.923333           0.654128   

     sin_wind_dir_t-47  ...        present_time       forecast_time  \
280           -0.71467  ... 2020-07-14 16:00:00 2020-07-14 22:00:00   

     speed_forecast  temp_forecast  rad_forecast  precip_forecast  \
280             4.6           23.1           0.0              0.0   

     cos_wind_dir_forecast  sin_wind_dir_forecast  cos_day  cos_hour  
280                0.75471               0.656059   0.9711      -0.5  

[1 rows x 353 columns]

In [30]:
#predict next 48 hours
predict_list = np.arange(1,49,1)
result = pd.DataFrame(columns={})
for pred in predict_list:
    x_test = prepare_x_test(measurement, forecast, 48, int(pred))
    #drop timestamp columns and change to array format
    f_date = x_test['forecast_time'].iloc[0]
    p_date = x_test['present_time'].iloc[0]
    x_test = x_test.drop(['present_time','forecast_time'], axis=1)

    #check missing features 
    xgb1 = pickle.load(open('../trained_models/speed_t_'+str(pred), 'rb'))
    model_features = xgb1.get_booster().feature_names
    missing_features = set(model_features).difference(x_test.columns)
    if len(missing_features)>=1:
        print('model not run because some features are missing:', missing_features)
        break 
    else:
        #re-order features 
        x_test = x_test[model_features]
        
    #read 3 models for speed, cos_wind, sin_wind
    xgb1= pickle.load(open('../trained_models/speed_t_'+str(pred), 'rb'))
    xgb2 = pickle.load(open('../trained_models/cos_wind_dir_t_'+str(pred), 'rb'))
    xgb3 = pickle.load(open('../trained_models/sin_wind_dir_t_'+str(pred), 'rb'))

    #predict 
    speed = xgb1.predict(x_test)[-1]
    cos_wind = xgb2.predict(x_test)[-1]
    sin_wind = xgb3.predict(x_test)[-1]
    
    #calculate wind_dir
    wind_dir = utils.get_angle_in_degree(cos_wind, sin_wind)

    #record result
    result = result.append({'forecast_time':f_date,
                            'present_time':p_date,
                            'speed':speed,
                            'cos_wind_dir':cos_wind,
                           'sin_wind_dir':sin_wind,
                           'wind_dir':wind_dir}, ignore_index=True)


In [31]:
result

cos_wind_dir       forecast_time        present_time  sin_wind_dir  \
0       0.844020 2020-07-14 17:00:00 2020-07-14 16:00:00     -0.441228   
1       0.897209 2020-07-14 18:00:00 2020-07-14 16:00:00     -0.243845   
2       0.887204 2020-07-14 19:00:00 2020-07-14 16:00:00      0.056082   
3       0.867661 2020-07-14 20:00:00 2020-07-14 16:00:00      0.329359   
4       0.796441 2020-07-14 21:00:00 2020-07-14 16:00:00      0.604113   
5       0.718218 2020-07-14 22:00:00 2020-07-14 16:00:00      0.618791   
6       0.683517 2020-07-14 23:00:00 2020-07-14 16:00:00      0.638492   
7       0.489306 2020-07-15 00:00:00 2020-07-14 16:00:00      0.763767   
8       0.688048 2020-07-15 01:00:00 2020-07-14 16:00:00      0.839634   
9       0.691381 2020-07-15 02:00:00 2020-07-14 16:00:00      0.697644   
10      0.654516 2020-07-15 03:00:00 2020-07-14 16:00:00      0.818058   
11      0.591090 2020-07-15 04:00:00 2020-07-14 16:00:00      0.835549   
12      0.551272 2020-07-15 05:00:00 2020-07-14 16:00:00      0.592261   
13      0.581036 2020-07-15 06:00:00 2020-07-14 16:00:00      0.801114   
14      0.524195 2020-07-15 07:00:00 2020-07-14 16:00:00      0.798282   
15      0.539059 2020-07-15 08:00:00 2020-07-14 16:00:00      0.796376   
16      0.551917 2020-07-15 09:00:00 2020-07-14 16:00:00      0.795032   
17      0.645015 2020-07-15 10:00:00 2020-07-14 16:00:00      0.704034   
18      0.631126 2020-07-15 11:00:00 2020-07-14 16:00:00      0.697406   
19      0.738388 2020-07-15 12:00:00 2020-07-14 16:00:00      0.192107   
20      0.814955 2020-07-15 13:00:00 2020-07-14 16:00:00     -0.541260   
21      0.763835 2020-07-15 14:00:00 2020-07-14 16:00:00     -0.618807   
22      0.831467 2020-07-15 15:00:00 2020-07-14 16:00:00     -0.547803   
23      0.878320 2020-07-15 16:00:00 2020-07-14 16:00:00     -0.581506   
24      0.774232 2020-07-15 17:00:00 2020-07-14 16:00:00     -0.364919   
25      0.874451 2020-07-15 18:00:00 2020-07-14 16:00:00     -0.067196   
26      0.766013 2020-07-15 19:00:00 2020-07-14 16:00:00      0.102740   
27      0.705268 2020-07-15 20:00:00 2020-07-14 16:00:00      0.350647   
28      0.666803 2020-07-15 21:00:00 2020-07-14 16:00:00      0.628611   
29      0.578299 2020-07-15 22:00:00 2020-07-14 16:00:00      0.556012   
30      0.402128 2020-07-15 23:00:00 2020-07-14 16:00:00      0.698242   
31      0.521588 2020-07-16 00:00:00 2020-07-14 16:00:00      0.716622   
32      0.514582 2020-07-16 01:00:00 2020-07-14 16:00:00      0.603665   
33      0.507237 2020-07-16 02:00:00 2020-07-14 16:00:00      0.392331   
34      0.588922 2020-07-16 03:00:00 2020-07-14 16:00:00      0.649776   
35      0.641657 2020-07-16 04:00:00 2020-07-14 16:00:00      0.692487   
36      0.582964 2020-07-16 05:00:00 2020-07-14 16:00:00      0.562848   
37      0.497755 2020-07-16 06:00:00 2020-07-14 16:00:00      0.372415   
38      0.478110 2020-07-16 07:00:00 2020-07-14 16:00:00      0.395461   
39      0.497064 2020-07-16 08:00:00 2020-07-14 16:00:00      0.443410   
40      0.460423 2020-07-16 09:00:00 2020-07-14 16:00:00      0.460500   
41      0.524228 2020-07-16 10:00:00 2020-07-14 16:00:00     -0.119049   
42      0.553910 2020-07-16 11:00:00 2020-07-14 16:00:00     -0.035643   
43      0.670052 2020-07-16 12:00:00 2020-07-14 16:00:00     -0.546250   
44      0.723595 2020-07-16 13:00:00 2020-07-14 16:00:00     -0.536212   
45      0.794808 2020-07-16 14:00:00 2020-07-14 16:00:00     -0.558260   
46      0.739807 2020-07-16 15:00:00 2020-07-14 16:00:00     -0.601162   
47      0.758736 2020-07-16 16:00:00 2020-07-14 16:00:00     -0.557717   

       speed    wind_dir  
0   4.952895  212.432948  
1   4.951971  206.206432  
2   4.772141   27.476031  
3   5.035214   29.812082  
4   5.064035   37.208429  
5   5.090146   44.092426  
6   4.246524   46.880940  
7   3.881823   60.705000  
8   3.733252   46.524233  
9   4.212215   46.260479  
10  4.333224   49.117028  
11  4.092558   53.765579  
12  4.057336   56.545681  

Hi Karim,

Please find my the updated notebook attached.\
I have made several changes to the code, and I will outline the changes made here: \
- crop out 7 hours of prediction data instead of 6
- when taking hourly average, now we are averaging to the 'right' instead of 'left' to reflect the hour reflect past hour data 
- xgb models are trained with column names. In addition, I've added several lines of code to compare model_features with testing_features, and the code will stop and raise an error flag of the missing features.

Models are all retrained to reflect the above amendments. I have pushed several directories to the github, and also sent the same files in a zip folder via Teams, including:
- trained_models: containing all the models in pickle files
- results: testing results (80% training / 20% testing data split) on the data set we have, to check performance after making the above 3 amendments. I have tested on max_depth=5, n_estimators = 100 and 200 respectively. Results are can be found in "xgboost_accuracy_in_48_depth_5_estim_100.csv" and "xgboost_accuracy_in_48_depth_5_estim_200.csv". I've compared this results with previously; now they are slightly worse (~3-4% decrease in accuracy), I think due to 1)croping out 1 more hour of forecast 2) shifting the hour to 1 hour later. Nevertheless, they are still good and consistent with accuracy at 83% for 3 hour prediction, 80% for 48 hour prediction.  

Please let me know if you have any problems with running the code, and good luck with the implementation process! 


Best,
Cynthia 